pkg/registry/core/service/ipallocator/controller/repairip.go

```diff
@@ -0,0 +1,539 @@
+			runtime.HandleError(fmt.Errorf("the ClusterIP %s for Service %s/%s is not a valid IP; please recreate Service", ip, svc.Name, svc.Namespace))
+			continue
+		}
+
+		family := getFamilyByIP(ip)
+		if network, ok := r.networkByFamily[family]; !ok {
+			// this service is using an IPFamily no longer configured on cluster
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotValid", "ClusterIPValidation", "Cluster IP %s(%s) is of ip family that is no longer configured on cluster; please recreate Service", ip, family)
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is of ip family that is no longer configured on cluster; please recreate Service", family, ip, svc.Name, svc.Namespace))
+			continue
+		} else if !network.Contains(ip) {
+			// ClusterIP is out of range
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPOutOfRange", "ClusterIPAllocation", "Cluster IP [%v]: %s is not within the configured Service CIDR %s; please recreate service", family, ip, network.String())
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is not within the service CIDR %s; please recreate", family, ip, svc.Name, svc.Namespace, network.String()))
+			continue
+		}
+
+		ipaddress, err := r.ipAddressLister.Get(ip.String())
+		if apierrors.IsNotFound(err) {
+			// ClusterIP doesn't seem to be allocated,
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotAllocated", "ClusterIPAllocation", "Cluster IP [%v]: %s is not allocated; repairing", family, ip)
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is not allocated; repairing", family, ip, svc.Name, svc.Namespace))
+			_, err := r.client.NetworkingV1alpha1().IPAddresses().Create(context.Background(), newIPAddress(ip.String(), family, svc), metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+			continue
+		}
+		if err != nil {
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "UnknownError", "ClusterIPAllocation", "Unable to allocate ClusterIP [%v]: %s due to an unknown error", family, ip)
+			return fmt.Errorf("unable to allocate ClusterIP [%v]: %s for Service %s/%s due to an unknown error, will retry later: %v", family, ip, svc.Name, svc.Namespace, err)
+		}
+		// IPAddress that belongs to a Service must reference the Service
+		if !reflect.DeepEqual(serviceToRef(svc), ipaddress.Spec.ParentRef) {
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotAllocated", "ClusterIPAllocation", "the ClusterIP [%v]: %s for Service %s/%s has a wrong reference; repairing", family, ip, svc.Namespace, svc.Name)
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s has a wrong reference %v; repairing", family, ip, svc.Name, svc.Namespace, ipaddress.Spec.ParentRef))
+			_, err := r.client.NetworkingV1alpha1().IPAddresses().Update(context.Background(), newIPAddress(ip.String(), family, svc), metav1.UpdateOptions{})
```

Comment:
What if it was owned by another service?

Reply:
oh, you are right, I should verify that doesn't exists Services with the same ClusterIP ... it can not happen ... unless there is a bug